File: leetcode/java/1448.count-good-nodes-in-binary-tree.java
class Solution {
    public int goodNodes(TreeNode root) {
        if (root == null) {
            return 0;
        }
        return 1 + goodNodes(root.left, root.val) + goodNodes(root.right, root.val);
    }

    public int goodNodes(TreeNode root, int m) {
        if (root == null)
            return 0;
        int isGood = root.val >= m ? 1 : 0;
        m = Math.max(m, root.val);
        return isGood + goodNodes(root.left, m) + goodNodes(root.right, m);
    }
}

File: leetcode/kt/1448.count-good-nodes-in-binary-tree.kt
class Solution {
    fun goodNodes(root: TreeNode?): Int {
        if (root == null) {
            return 0
        }
        return 1 + goodNodes(root.left, root.`val`) + goodNodes(root.right, root.`val`)
    }

    fun goodNodes(root: TreeNode?, _max: Int): Int {
        if (root == null) return 0
        val isGood = if (root.`val` >= _max) 1 else 0
        val _m = Math.max(_max, root.`val`)
        return isGood + goodNodes(root.left, _m) + goodNodes(root.right, _m)
    }

}
